src/utils/read-rows-resumption.ts

```diff
@@ -0,0 +1,267 @@
+    // Add an empty range to simplify the resumption logic.
+    if (this.rowKeys.length === 0 && this.ranges.length === 0) {
+      this.ranges.push({});
+    }
+  }
+
+  /**
+   * Gets the next readrows request.
+   *
+   * This function computes the next readRows request that will be sent to the
+   * server. Based on the last row key calculated by data already passed through
+   * the chunk transformer, the set of row keys and row ranges is calculated and
+   * updated. The calculated row keys and ranges are used along with other
+   * properties provided by the user like limits and filters to compute and
+   * return a request that will be used in the next read rows call.
+   *
+   * @return {protos.google.bigtable.v2.IReadRowsRequest} The request options
+   * for the next readrows request.
+   */
+  getResumeRequest(): protos.google.bigtable.v2.IReadRowsRequest {
+    const reqOpts = this
+      .tableStrategyInfo as google.bigtable.v2.IReadRowsRequest;
+
+    // Create the new reqOpts
+    reqOpts.rows = {};
+
+    // TODO: preprocess all the keys and ranges to Bytes
+    reqOpts.rows.rowKeys = this.rowKeys.map(
+      Mutation.convertToBytes
+    ) as {} as Uint8Array[];
+
+    reqOpts.rows.rowRanges = this.ranges.map(range =>
+      Filter.createRange(
+        range.start as BoundData,
+        range.end as BoundData,
+        'Key'
+      )
+    );
+
+    if (this.options.filter) {
+      reqOpts.filter = Filter.parse(this.options.filter);
+    }
+
+    if (this.hasLimit) {
+      reqOpts.rowsLimit = this.rowsLimit - this.rowsRead;
+    }
+    return reqOpts;
+  }
+
+  /**
+   * Decides if the client is going to retry a request.
+   *
+   * canResume contains the logic that will decide if the client will retry with
+   * another request when it receives an error. This logic is passed along to
+   * google-gax and used by google-gax to decide if the client should retry
+   * a request when google-gax receives an error. If canResume returns true then
+   * the client will retry with another request computed by getResumeRequest. If
+   * canResume request returns false then the error will bubble up from gax to
+   * the handwritten layer.
+   *
+   * @param {GoogleError} [error] The error that Google Gax receives.
+   * @return {boolean} True if the client will retry
+   */
+  canResume(error: GoogleError): boolean {
+    // First update the row keys and the row ranges based on the last row key.
+    const lastRowKey = this.chunkTransformer
+      ? this.chunkTransformer.lastRowKey
+      : '';
+    if (lastRowKey) {
+      spliceRanges(this.ranges, lastRowKey);
+      this.rowKeys = getRowKeys(this.rowKeys, lastRowKey);
+    }
+
+    // If all the row keys and ranges are read, end the stream
+    // and do not retry.
+    if (this.rowKeys.length === 0 && this.ranges.length === 0) {
+      return false;
+    }
+    // If there was a row limit in the original request and
+    // we've already read all the rows, end the stream and
```

Comment:
nit - possible rephrase - should this be something like "we've met/exceeded that limit" or "we've already read the number of rows in that limit"?

Reply:
I updated this comment. This was just [copy/paste](https://github.com/googleapis/nodejs-bigtable/blob/b2b2fdae61865a36efc8f1f077276053f41d992f/src/table.ts#L858), much of the code in `read-rows-resumption.ts` used to live in `table.ts`.